### PR TITLE
[libc++] Allow the use of extensions in the implementation

### PR DESCRIPTION
**Pull request by philnik777**
We've talked about allowing extensions on [discourse](https://discourse.llvm.org/t/rfc-use-language-extensions-from-future-standards-in-libc/71898/5) and in a libc++ monthly meeting and agreed to test it out in the LLVM 18 release. We've done that with the `tuple` constructor overload set (using conditional `explicit`). Since we haven't heard about any breakages, it seems safe to do. This patch enables the use of extension from later C++ standards inside the versioned `std` namespaces. This should be good enough, since almost all of our code is inside that namespace. This approach also avoids the use of extensions inside the test `std` suite. That part of the code base should stay clean, since it's a test suite that is also used by other vendors to test their implementations.

### CODE REVIEW

File: libcxx/include/__config
Comment (ldionne) [inline libcxx/include/__config:787]: Please add a description to this patch.

File: libcxx/include/__config
Comment (ldionne) [inline libcxx/include/__config:901]: I would suggest defining a shorthand for these to avoid cluttering the namespace definition too much.

Comment (ldionne) [reply libcxx/include/__config:901]: As discussed, this is optional.

**Review by ldionne — APPROVED** [review]
We discussed it multiple times and this is the end goal we've been driving towards. I think this has the potential to greatly simplify our implementation in places.

However, let's be reasonable and only use these extensions when they are actually useful -- in particular we shouldn't start reimplementing stuff using extensions just because we can, we should only do it when there's a tangible benefit. Let's use our judgement.

**Comment by var-const** [comment]
This is awesome! Is there a list of which extensions are available in which language modes -- or perhaps we should create one? It doesn't have to be exhaustive -- just listing some of the major, most clearly useful extensions with the minimum language standard where it can be used would be really useful, IMO (we can also have a checkbox saying whether there's already precedent for using the extension in the codebase, it might help rolling this out more gradually).

**Comment by mordante** [comment]
> We've talked about allowing extensions on [discourse](https://discourse.llvm.org/t/rfc-use-language-extensions-from-future-standards-in-libc/71898/5) and in a libc++ monthly meeting and agreed to test it out in the LLVM 18 release. We've done that with the `tuple` constructor overload set (using conditional `explicit`). Since we haven't heard about any breakages, it seems safe to do. 

Since we don't have a LLVM-18 release yet and presumably not a lot of usage except for users HEAD I feel this conclusion is premature. I  really like to wait until after the release and see whether we get breakage before moving forward with this patch.

**Comment by philnik777** [comment]
> > We've talked about allowing extensions on [discourse](https://discourse.llvm.org/t/rfc-use-language-extensions-from-future-standards-in-libc/71898/5) and in a libc++ monthly meeting and agreed to test it out in the LLVM 18 release. We've done that with the `tuple` constructor overload set (using conditional `explicit`). Since we haven't heard about any breakages, it seems safe to do.
> 
> Since we don't have a LLVM-18 release yet and presumably not a lot of usage except for users HEAD I feel this conclusion is premature. I really like to wait until after the release and see whether we get breakage before moving forward with this patch.

We've had this in trunk for months now and nobody said anything. rc1 has been out for a week now and nobody said anything. If we caused any non-trivial breakage I'm pretty sure we would have heard about it by now. Everybody wants to start simplifying the code, and I don't think "but maybe" should stop us from doing so by now.

**Comment by ldionne** [comment]
@mordante @philnik777 I'd be fine to wait until after we have `18.1.0` released (maybe 1-2 weeks after that) before we actually merge this. At this point we've been waiting for long enough that a few weeks is not going to make much of a difference.

For what it's worth, I have been deploying LLVM 18 on a large code base internally and didn't see any issues around the use of these extensions, so I am not too concerned at this point. But I agree that a full point release would make sense before we actually merge this.

**Comment by mordante** [comment]
> @mordante @philnik777 I'd be fine to wait until after we have `18.1.0` released (maybe 1-2 weeks after that) before we actually merge this. At this point we've been waiting for long enough that a few weeks is not going to make much of a difference.
> 
> For what it's worth, I have been deploying LLVM 18 on a large code base internally and didn't see any issues around the use of these extensions, so I am not too concerned at this point. But I agree that a full point release would make sense before we actually merge this.

Thanks for testing this. My main concern was never with users who we know about, but about the users we don't know about.

**Comment by EricWF** [comment]
Could you adjust the commit message to better describe which extensions we're enabling when?

**Comment by sam-mccall** [comment]
We're seeing major compile-time regressions (50%+) from this change in normal-looking code that uses libc++, using clang from head.

My best guess is that we're now adding lots of diagnostics-overridden regions to most TUs, and querying them (when clang checks: should we emit diagnostic X) is inefficient. I will dig a bit into this.

If this is not related to our use of modules (I need to check) then I think we should revert this and hopefully find a cheaper way. WDYT?
If it is modules-related, then it seems more complicated and will depend on the details.

**Comment by sam-mccall** [comment]
(Sorry this is coming a week after the patch landed: internally we do basic correctness testing within days when we pull in the code, but only do performance testing for releases)

**Comment by EricWF** [comment]
If it is because of the diagnostic overrides, we should be able to go back to simply using the pragma system header.

The diagnostic disabling is only really for our test suite (and people who enable -Wsystem-header, but they asking to see diagnostics). We should just be able to disable the warning entirely in the test suite. IDK Why we have it on when we allow the use of extensions. 

@philnik777 Can you say why you chose to go this route?

**Comment by sam-mccall** [comment]
Thanks! If we can avoid this entirely that seems ideal for everyone.

We only see the regression in our modular builds. (Which unfortunately is our main mode).
We use clang header modules, but I'd guess this isn't specific to that mode.
libc++ has modules buildbots right? I'll see if I can scrape some timings out of those, 50% might be possible to see through noise.

**Comment by EricWF** [comment]
Is there a way to observe the issue or at least the structure that causes
the issue in the AST?

**Comment by EricWF** [comment]
Actually, sorry I need to step back from this review.

**Comment by sam-mccall** [comment]
No worries @EricWF, thanks for your help.

The regression is time spent in ASTReader::ReadPragmaDiagnosticMappings, which explains why it's modules-specific.

image

here's the baseline

image

and a ridiculously-wide screenshot of just that function

image

(sorry I don't have a reasonable way to share the underlying data)

**Comment by sam-mccall** [comment]
My approximate understanding is that any query of diagnostic overrides ends up loading all the mappings (which is a bunch of work), and in order to understand the region they apply to it has to page in source-level information (since they are source-location based).
Maybe this can be fixed eventually in clang, but we need some solution meanwhile.

@philnik777 thanks for the context!

> I'm pretty sure I've checked compile times before and after and didn't see any significant regression

(I would assume this is using libc++ textually rather than as modules, but just want to check)

> We're already depending on this in parts of the code base. IMO we should fix forward instead if this is actually a problem with this approach.

Can you tell me how to check how much would break? (I know how to test llvm but not libcxx specifically).
As a short-term fix, we could enable extensions for just those regions/files until we find a way to do it globally that scales well.

> I've done it this way to avoid introducing extensions in the test suite, since that shouldn't be using extensions as much as possible as a conformance test suite and it's used by other implementations

It seems nice to enforce this, but if we can't revert and there isn't another obvious fix-forward, is it possible to live without this enforcement? Or move the pragmas out of libc++ proper and into the testsuite?

**Comment by sam-mccall** [comment]
Yes, that solves the problems I know of.

Would need other libc++ maintainers to weigh in on what the right tradeoff is here.

**Pull request merged**
